diff --git a/CMakeLists.txt b/CMakeLists.txt
index 6bd3886..4432a60 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -13,8 +13,8 @@ set(CMAKE_STANDARD 14)
 if(MSVC)
 	set(CMAKE_CXX_COMPILER $ENV{VCToolsInstallDir}/bin/Hostx64/x64/cl.exe)
 	set(CMAKE_CXX_LINK_EXECUTABLE $ENV{VCToolsInstallDir}/bin/Hostx64/x64/link.exe)
-else()
-	set(CMAKE_CXX_COMPILER g++)
+#else()
+	#set(CMAKE_CXX_COMPILER g++)
 endif()
 
 project(diskann)
diff --git a/include/aux_utils.h b/include/aux_utils.h
index f980b7f..83a3633 100644
--- a/include/aux_utils.h
+++ b/include/aux_utils.h
@@ -29,7 +29,7 @@ typedef int FileHandle;
 #include "common_includes.h"
 #include "utils.h"
 #include "windows_customizations.h"
-#include "gperftools/malloc_extension.h"
+//#include "gperftools/malloc_extension.h"
 
 namespace diskann {
   const size_t   TRAINING_SET_SIZE = 100000;
diff --git a/include/distance.h b/include/distance.h
index 086c428..58eb6bb 100644
--- a/include/distance.h
+++ b/include/distance.h
@@ -1,6 +1,6 @@
 #pragma once
 
-#include <utils.h>
+#include "utils.h"
 #ifdef _WINDOWS
 #include <immintrin.h>
 #include <smmintrin.h>
@@ -10,7 +10,7 @@
 #include <immintrin.h>
 #endif
 
-#include <cosine_similarity.h>
+#include "cosine_similarity.h"
 #include <iostream>
 
 namespace {
diff --git a/include/index.h b/include/index.h
index fdcac5f..650baad 100644
--- a/include/index.h
+++ b/include/index.h
@@ -55,7 +55,8 @@ namespace diskann {
     DISKANN_DLLEXPORT std::pair<uint32_t, uint32_t> search(const T *      query,
                                                            const size_t   K,
                                                            const unsigned L,
-                                                           unsigned *indices);
+                                                           unsigned *indices,
+                                                           float *   distances);
 
     DISKANN_DLLEXPORT std::pair<uint32_t, uint32_t> search(
         const T *query, const uint64_t K, const unsigned L,
@@ -101,7 +102,14 @@ namespace diskann {
     DISKANN_DLLEXPORT void optimize_graph();
 
     DISKANN_DLLEXPORT void search_with_opt_graph(const T *query, size_t K,
-                                                 size_t L, unsigned *indices);
+                                                 size_t L, unsigned *indices,
+                                                 float *distances);
+
+    DISKANN_DLLEXPORT size_t get_dimension() const { return _dim; }
+
+    DISKANN_DLLEXPORT size_t get_vector_number() const { return _nd; }
+
+    DISKANN_DLLEXPORT const T* get_vector(int id) const { return _data + _aligned_dim * (size_t) id; }
 
     /*  Internals of the library */
    protected:
diff --git a/include/utils.h b/include/utils.h
index 9bf2e14..050e1c1 100644
--- a/include/utils.h
+++ b/include/utils.h
@@ -341,7 +341,7 @@ namespace diskann {
       dists = new float[npts * dim];
       reader.read((char*) dists, npts * dim * sizeof(float));
     }
-    float min_dist = std::numeric_limits<float>::max();
+    float min_dist = (std::numeric_limits<float>::max)();
     float max_dist = 0;
     groundtruth.resize(npts);
     for (_u32 i = 0; i < npts; i++) {
diff --git a/src/aux_utils.cpp b/src/aux_utils.cpp
index f143b88..028d839 100644
--- a/src/aux_utils.cpp
+++ b/src/aux_utils.cpp
@@ -808,7 +808,7 @@ namespace diskann {
     delete[] train_data;
 
     train_data = nullptr;
-    MallocExtension::instance()->ReleaseFreeMemory();
+    //MallocExtension::instance()->ReleaseFreeMemory();
 
     diskann::build_merged_vamana_index<T>(
         data_file_to_use.c_str(), diskann::Metric::L2, L, R, p_val,
diff --git a/src/index.cpp b/src/index.cpp
index 38b58da..56de6c4 100644
--- a/src/index.cpp
+++ b/src/index.cpp
@@ -1029,7 +1029,14 @@ namespace diskann {
   std::pair<uint32_t, uint32_t> Index<T, TagT>::search(const T *      query,
                                                        const size_t   K,
                                                        const unsigned L,
-                                                       unsigned *     indices) {
+                                                       unsigned *     indices,
+                                                       float *distances) {
+    T *    aligned_query;
+    size_t allocSize = _aligned_dim * sizeof(T);
+    alloc_aligned(((void **) &aligned_query), allocSize, 8 * sizeof(T));
+    memset(aligned_query, 0, _aligned_dim * sizeof(T));
+    memcpy(aligned_query, query, _dim * sizeof(T));
+
     std::vector<unsigned>    init_ids;
     tsl::robin_set<unsigned> visited(10 * L);
     std::vector<Neighbor>    best_L_nodes, expanded_nodes_info;
@@ -1039,12 +1046,14 @@ namespace diskann {
       init_ids.emplace_back(_ep);
     }
     auto retval =
-        iterate_to_fixed_point(query, L, init_ids, expanded_nodes_info,
+        iterate_to_fixed_point(aligned_query, L, init_ids, expanded_nodes_info,
                                expanded_nodes_ids, best_L_nodes);
+    aligned_free((void *)aligned_query);
 
     size_t pos = 0;
     for (auto it : best_L_nodes) {
       indices[pos] = it.id;
+      distances[pos] = it.distance;
       pos++;
       if (pos == K)
         break;
@@ -1086,7 +1095,8 @@ namespace diskann {
       unsigned *indices_buffer) {
     const bool alloc = indices_buffer == NULL;
     auto       indices = alloc ? new unsigned[K] : indices_buffer;
-    auto       ret = search(query, K, L, indices);
+    auto       dists = new float[K];
+    auto       ret = search(query, K, L, indices, dists);
     for (int i = 0; i < (int) K; ++i)
       tags[i] = _location_to_tag[indices[i]];
     if (alloc)
@@ -1121,7 +1131,8 @@ namespace diskann {
 
   template<typename T, typename TagT>
   void Index<T, TagT>::search_with_opt_graph(const T *query, size_t K, size_t L,
-                                             unsigned *indices) {
+                                             unsigned *indices,
+                                             float *   distances) {
     DistanceFastL2<T> *dist_fast = (DistanceFastL2<T> *) _distance;
 
     std::vector<Neighbor> retset(L + 1);
@@ -1215,6 +1226,7 @@ namespace diskann {
     }
     for (size_t i = 0; i < K; i++) {
       indices[i] = retset[i].id;
+      distances[i] = retset[i].distance;
     }
   }
 
diff --git a/tests/search_memory_index.cpp b/tests/search_memory_index.cpp
index e9951eb..45f8ce1 100644
--- a/tests/search_memory_index.cpp
+++ b/tests/search_memory_index.cpp
@@ -124,10 +124,10 @@ int search_memory_index(int argc, char** argv) {
       if (metric == diskann::FAST_L2) {
         index.search_with_opt_graph(
             query + i * query_aligned_dim, recall_at, L,
-            query_result_ids[test_id].data() + i * recall_at);
+            query_result_ids[test_id].data() + i * recall_at, query_result_dists[test_id].data() + i * recall_at);
       } else {
         index.search(query + i * query_aligned_dim, recall_at, L,
-                     query_result_ids[test_id].data() + i * recall_at);
+                     query_result_ids[test_id].data() + i * recall_at, query_result_dists[test_id].data() + i * recall_at);
       }
       auto qe = std::chrono::high_resolution_clock::now();
       std::chrono::duration<double> diff = qe - qs;
